{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "f4811e4e-cd7e-4cb2-a209-6eade0ec0537",
   "metadata": {},
   "source": [
    "# Microbiome Relative Abundance Data Cleanup\n",
    "## Anna Lapteva\n",
    "### April 10, 2024\n",
    "\n",
    "All the data used to generate the final .CSV document is sourced from [GMRepo](https://gmrepo.humangut.info/home) database. Specifically, all run IDs in the database linked to each project for each phenotype (Autism Spectrum Disorder (ASD), Alzheimer's Disease (AD), Parkinson's Disease (PD), Multiple Sclerosis (MS), and epilepsy) were saved as a spreadsheet, individually opened, and downloaded. Each downloaded .txt file is then cleaned and appended to a DataFrame in this code. The following projects, which perform 16S sequencing of fecal samples using Illumina tools, were used for data collection:\n",
    "- ASD: PRJNA578223\n",
    "- AD: PRJNA633959\n",
    "- PD: PRJEB17784\n",
    "- MS: PRJNA450340\n",
    "\n",
    "Let's start by importing relevant packages."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "e0a3634f-8c44-4c86-835f-7e4d629c483e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Relevant string operations\n",
    "import string\n",
    "\n",
    "# Data handling\n",
    "import glob\n",
    "import os\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "10d040bb-998a-4b7d-9e07-ed228e7b8c3a",
   "metadata": {},
   "source": [
    "We will use the `glob` module, which allows us to get all our files of interest with the same extension for convenience."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "f87d1734-e18c-48cd-88a1-681ccaee3d2c",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>scientific_name</th>\n",
       "      <th>Akkermansia</th>\n",
       "      <th>Bacteroides</th>\n",
       "      <th>Bifidobacterium</th>\n",
       "      <th>Butyricicoccus</th>\n",
       "      <th>Clostridium</th>\n",
       "      <th>Collinsella</th>\n",
       "      <th>Coprococcus</th>\n",
       "      <th>Enterococcus</th>\n",
       "      <th>Faecalibacterium</th>\n",
       "      <th>Gemmiger</th>\n",
       "      <th>...</th>\n",
       "      <th>Endozoicomonas</th>\n",
       "      <th>Hydrocarboniphaga</th>\n",
       "      <th>Brucella</th>\n",
       "      <th>Jannaschia</th>\n",
       "      <th>Labrys</th>\n",
       "      <th>Parvibaculum</th>\n",
       "      <th>Rudaea</th>\n",
       "      <th>Sphingopyxis</th>\n",
       "      <th>Serratia</th>\n",
       "      <th>Actinobaculum</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2.843730</td>\n",
       "      <td>18.38760</td>\n",
       "      <td>4.527880</td>\n",
       "      <td>0.193263</td>\n",
       "      <td>4.942020</td>\n",
       "      <td>2.098290</td>\n",
       "      <td>26.339000</td>\n",
       "      <td>0.193263</td>\n",
       "      <td>10.215400</td>\n",
       "      <td>4.224190</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>25.03320</td>\n",
       "      <td>2.251980</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>3.221140</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.560300</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.117950</td>\n",
       "      <td>3.313630</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.096056</td>\n",
       "      <td>8.59078</td>\n",
       "      <td>0.285050</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>4.491250</td>\n",
       "      <td>2.173740</td>\n",
       "      <td>1.911140</td>\n",
       "      <td>0.003119</td>\n",
       "      <td>4.920380</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.000710</td>\n",
       "      <td>5.34432</td>\n",
       "      <td>0.460558</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.048256</td>\n",
       "      <td>0.042579</td>\n",
       "      <td>0.001419</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.386755</td>\n",
       "      <td>0.009225</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.001890</td>\n",
       "      <td>6.94195</td>\n",
       "      <td>0.059074</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.034736</td>\n",
       "      <td>0.000236</td>\n",
       "      <td>0.107043</td>\n",
       "      <td>0.008979</td>\n",
       "      <td>0.252839</td>\n",
       "      <td>0.014414</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 892 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "scientific_name  Akkermansia  Bacteroides  Bifidobacterium  Butyricicoccus  \\\n",
       "0                   2.843730     18.38760         4.527880        0.193263   \n",
       "1                   0.000000     25.03320         2.251980        0.000000   \n",
       "2                   0.096056      8.59078         0.285050        0.000000   \n",
       "3                   0.000710      5.34432         0.460558        0.000000   \n",
       "4                   0.001890      6.94195         0.059074        0.000000   \n",
       "\n",
       "scientific_name  Clostridium  Collinsella  Coprococcus  Enterococcus  \\\n",
       "0                   4.942020     2.098290    26.339000      0.193263   \n",
       "1                   3.221140     0.000000     1.560300      0.000000   \n",
       "2                   4.491250     2.173740     1.911140      0.003119   \n",
       "3                   0.048256     0.042579     0.001419      0.000000   \n",
       "4                   0.034736     0.000236     0.107043      0.008979   \n",
       "\n",
       "scientific_name  Faecalibacterium  Gemmiger  ...  Endozoicomonas  \\\n",
       "0                       10.215400  4.224190  ...             0.0   \n",
       "1                        1.117950  3.313630  ...             0.0   \n",
       "2                        4.920380  0.000000  ...             0.0   \n",
       "3                        0.386755  0.009225  ...             0.0   \n",
       "4                        0.252839  0.014414  ...             0.0   \n",
       "\n",
       "scientific_name  Hydrocarboniphaga  Brucella  Jannaschia  Labrys Parvibaculum  \\\n",
       "0                              0.0       0.0         0.0     0.0          0.0   \n",
       "1                              0.0       0.0         0.0     0.0          0.0   \n",
       "2                              0.0       0.0         0.0     0.0          0.0   \n",
       "3                              0.0       0.0         0.0     0.0          0.0   \n",
       "4                              0.0       0.0         0.0     0.0          0.0   \n",
       "\n",
       "scientific_name  Rudaea  Sphingopyxis  Serratia  Actinobaculum  \n",
       "0                   0.0           0.0       0.0            0.0  \n",
       "1                   0.0           0.0       0.0            0.0  \n",
       "2                   0.0           0.0       0.0            0.0  \n",
       "3                   0.0           0.0       0.0            0.0  \n",
       "4                   0.0           0.0       0.0            0.0  \n",
       "\n",
       "[5 rows x 892 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Specify directories holding data\n",
    "data_dir = ['../Data/LargeData/ASD', '../Data/LargeData/AD', '../Data/LargeData/PD', '../Data/LargeData/MS']\n",
    "            \n",
    "# Initialize DataFrame for concatenation steps\n",
    "df = pd.DataFrame()\n",
    "\n",
    "# Iterate through each of the directories\n",
    "for direc in data_dir:\n",
    "    \n",
    "    # Glob string for all .txt files, which have relative abundance data\n",
    "    file_glob = os.path.join(direc, '*.txt')\n",
    "    \n",
    "    # Get list of files in direc\n",
    "    file_list = glob.glob(file_glob)\n",
    "\n",
    "    # Iterate through all files in list\n",
    "    for file in file_list:\n",
    "\n",
    "        # Initialize skip_row_id to be sufficiently large before the loop starts\n",
    "        skip_row_id = len(file)\n",
    "\n",
    "        # List to store modified lines\n",
    "        modified_lines = []\n",
    "        \n",
    "        # Iterate through all lines in file\n",
    "        with open(file, 'r') as f:\n",
    "            # Find where relative abundance data starts\n",
    "            for row_number, line in enumerate(f, start=1):\n",
    "                if \"ncbi_taxon_id\" in line:\n",
    "                    skip_row_id = row_number - 1\n",
    "        \n",
    "                # Ensure column entries are one \"word\"\n",
    "                if row_number > skip_row_id:  # Ensure lines before are not edited\n",
    "                    modified_line = ''\n",
    "                    prev_char = None\n",
    "                    for char in line:\n",
    "                        if char == ' ' and prev_char in string.ascii_lowercase + '.':\n",
    "                            modified_line += '_'\n",
    "                        else:\n",
    "                            modified_line += char\n",
    "                        prev_char = char\n",
    "                    modified_lines.append(modified_line)\n",
    "        \n",
    "        # Write modified lines back to file\n",
    "        with open(file, 'w') as f:\n",
    "            f.writelines(modified_lines)\n",
    "\n",
    "        # Create DataFrame from file, removing filler info\n",
    "        data = pd.read_csv(file, sep='\\s+')\n",
    "\n",
    "        # Add individual index for reshaping\n",
    "        data[\"individual\"] = 0\n",
    "\n",
    "        # Reshape data\n",
    "        data = data[[\"relative_abundance\", \"scientific_name\", \"individual\"]].pivot_table(columns=\"scientific_name\", values=\"relative_abundance\", index=\"individual\", aggfunc=\"sum\").reset_index()\n",
    "        data = data.drop(\"individual\", axis=1)\n",
    "\n",
    "        # Add label\n",
    "        data[\"Diagnosis\"] = [direc[direc.rfind('/') + 1:]] * len(data)\n",
    "\n",
    "        # Update DataFrame\n",
    "        df = pd.concat([df, data], ignore_index=True).fillna(0)\n",
    "\n",
    "# Check out df\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5efe70b8-1d3b-4019-88e4-9ed322c94f28",
   "metadata": {},
   "source": [
    "Finally, let's export the data as a CSV!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "8fd4dc71-576a-4938-a1f6-5fb288868e8e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Export to CSV\n",
    "df.to_csv(\"../Data/20440_cleaned_data_large.csv\", index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
